lib/src/model/game/chat_controller.dart

```diff
@@ -0,0 +1,118 @@
+import 'dart:async';
+
+import 'package:fast_immutable_collections/fast_immutable_collections.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:lichess_mobile/src/model/auth/auth_socket.dart';
+import 'package:lichess_mobile/src/model/common/socket.dart';
+import 'package:lichess_mobile/src/model/game/game_socket_events.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'chat_controller.freezed.dart';
+part 'chat_controller.g.dart';
+
+@riverpod
+class ChatController extends _$ChatController {
+  StreamSubscription<SocketEvent>? _socketSubscription;
+
+  int? _socketEventVersion;
+
+  AuthSocket get _socket => ref.read(authSocketProvider);
+
+  @override
+  Future<ChatState> build() {
+    _socketEventVersion = 0;
+    final stream = _socket.stream;
+    stream.listen(_handleSocketTopic);
+
+    ref.onDispose(() {
+      _socketSubscription?.cancel();
+    });
+
+    return stream.firstWhere((e) => e.topic == 'full').then(
```

Comment:
in order to be generic, the chat controller shouldn't listen to the `full` event which is only related to "game" socket.
It should be easy to replace this by a setter:
```
void setMessages(IList<Message> messages) {
   state = ChatState(
     messages: messages,
     unreadMessages: 0,
   );
}
```

the `build` method does not return a Future anymore, it will even simplify things